From f41f48e8b80f2c03d0ba113495d87fea83f8e38f Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 05 Jan 2021 18:48:54 +0200
Subject: [PATCH] [SW-20719] gaudi: pll: add hbm related plls init

Add NIC and HBM PLLs initialization, so we can run the HBM init code
which requires those PLLs to be initialized.
The NIC PLLs are set to their regular frequency - 833 MHz.
The HBM PLLs are set to 800MHz - the safe frequency for all device
types.

Change-Id: I53335a11edad5088bc7d9e74de0fa38892a7a60b
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/Makefile b/gaudi/Makefile
index e19dda2..beda2c6 100644
--- a/gaudi/Makefile
+++ b/gaudi/Makefile
@@ -1,6 +1,7 @@
 FW_SIZE := 0x4000
 PPBOOT_MAX_SIZE := 16384
 
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/pll.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/security.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/reset.c)
 
diff --git a/gaudi/pll.c b/gaudi/pll.c
new file mode 100644
index 0000000..241433c
--- /dev/null
+++ b/gaudi/pll.c
@@ -0,0 +1,58 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2016-2021 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+#include <stddef.h>
+
+#include <gaudi_blocks.h>
+#include <preboot.h>
+#include <pll.h>
+
+#include "common.h"
+
+static struct hl_pll_cfg gaudi_pll_cfg[] = {
+	{
+		.name = "NIC0", .base = mmNIC0_PLL_BASE & ~HL_CFG_BASE,
+		.nr = 24, .nf = 1665, .od = 3, .nb = 832, /* 833 MHz */
+		.div_factor = { 0, 4, 8, -1 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_REF_CLK
+		},
+	}, {
+		.name = "NIC1", .base = mmNIC1_PLL_BASE & ~HL_CFG_BASE,
+		.nr = 24, .nf = 1665, .od = 3, .nb = 832, /* 833 MHz */
+		.div_factor = { 0, 4, 8, -1 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_REF_CLK
+		},
+	}, {
+		.name = "HBM", .base = mmPSOC_HBM_PLL_BASE & ~HL_CFG_BASE,
+		.nr = 0, .nf = 63, .od = 3, .nb = 31, /* 800 MHz */
+		.div_factor = { 0, 1, -1, -1 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_REF_CLK,
+			     HL_PLL_REF_CLK
+		},
+	}, {
+		.name = "NIC2_HBM", .base = mmNIC2_HBM_PLL_BASE & ~HL_CFG_BASE,
+		.nr = 0, .nf = 63, .od = 3, .nb = 31, /* 800 MHz */
+		.div_factor = { 0, 1, -1, -1 },
+		.div_sel = { HL_PLL_PLL_CLK,
+			     HL_PLL_DIV_PLL,
+			     HL_PLL_REF_CLK,
+			     HL_PLL_REF_CLK
+		},
+	},
+};
+
+void hl_dev_pll_init(void)
+{
+	hl_pll_init(gaudi_pll_cfg, ARRAY_SIZE(gaudi_pll_cfg));
+}
